### PR TITLE
fix(ca9): Extract opinions from  dynamodb tables 

### LINKED ISSUE #2120: `ca9_p` / `ca9_u` scrapers missing ~15% of opinions
`ca9_p` and `ca9_u` read the court's RSS feeds:

- https://www.ca9.uscourts.gov/decisions/opinions/index.xml
- https://www.ca9.uscourts.gov/decisions/memoranda/index.xml

Both feeds hold **25 items** and carry no upload signal — `<pubDate>` is the filing date. Comparing the court's own data against CourtListener shows we are losing a substantial fraction of both.

This is the same root cause as #2111 (`ca9` oral arguments), fixed for oral arguments in #2119. Filing one issue for both scrapers since `ca9_u.Site` subclasses `ca9_p.Site` and they share the defect.

## Impact

Opinions and memoranda **filed 2026-05-18 through 2026-05-31**:

| | count |
|---|---|
| in the court's `opinions` + `memoranda` tables | 180 |
| in CourtListener ([link][cl], published + unpublished) | 154 |
| **missing** | **27 (15%)** |

[cl]: https://www.courtlistener.com/?q=&type=o&court=ca9&filed_after=05%2F18%2F2026&filed_before=05%2F31%2F2026&stat_Published=on&stat_Unpublished=on

Missing by filing date: 05-21 → 4, 05-22 → 2, 05-26 → 2, 05-27 → 9, 05-29 → 10.

There is no Sentry issue for this: nothing errors. The scraper reports success every run and silently stops partway down the feed.

## Root cause

**26 of the 27 missing items were posted the same day they were filed.** Late posting is not the main driver.

The court publishes in a trickle through the morning — on 2026-05-27, 40 items between 08:35 and 13:22, roughly one every five minutes. Every item filed that day carries the **same** `<pubDate>`, so `AbstractSite._date_sort` falls through to its next tiebreaker and orders them **by case name**.

So on each hourly run, the items already ingested are scattered alphabetically among the new ones. `DupChecker.press_on` hits 5 consecutive duplicates, raises `ConsecutiveDuplicatesError`, `scrape_court` breaks, and everything below that point is skipped. Later postings then push those skipped items off the 25-item feed, and they are gone permanently.

Two aggravating factors that the oral argument scraper did not have:

1. **The feed is a hard 25-item window.** For opinions that is ~14 days of cover; for memoranda it is **2 days**, and on a 40-filing day it turns over in hours. An item skipped once is not skipped until next run, it is lost.
2. **Neither scraper supports backscraping.** No `_download_backwards`, no `make_backscrape_iterable`, no `first_opinion_date`, and the feed URLs accept no date or page parameters. Nothing can recover a missed item.

<details><summary>The 27 missing opinions, with the court's posting timestamps</summary>

```
filed        posted                lag   docket      case
2026-05-21   2026-05-21 09:20:04   0d    21-1276     MANZANARES-VALENTE V. BLANCHE
2026-05-21   2026-05-21 09:30:04   0d    22-231      TOMAS AGUSTIN V. BLANCHE
2026-05-21   2026-05-21 10:20:05   0d    25-1267     USA V. ROGERS
2026-05-21   2026-05-21 10:35:04   0d    25-6945     HARADA FAMILY DENTAL CARE, PC V. KAPITUS
2026-05-22   2026-05-22 09:25:04   0d    25-7765     HARTY, ET AL. V. COMMISSIONER OF INTERNAL REVENUE
2026-05-22   2026-05-22 10:05:05   0d    25-6274     MORALES-ITZEP V. BLANCHE
2026-05-26   2026-05-26 09:25:04   0d    24-2759     USA V. RUIZ-HERNANDEZ
2026-05-26   2026-05-26 11:10:05   0d    25-3456     RUIZ-ACOSTA, ET AL. V. BLANCHE
2026-05-27   2026-05-27 09:10:04   0d    23-595      USA V. HACKETT
2026-05-27   2026-05-27 09:40:04   0d    24-3621     HAMMETT V. STERN, ET AL.
2026-05-27   2026-05-27 11:05:04   0d    25-3846     ROJAS-ARROYO, ET AL. V. BLANCHE
2026-05-27   2026-05-27 11:10:06   0d    24-2682     SULLIVAN V. STATE OF MONTANA, ET AL.
2026-05-27   2026-05-27 11:25:06   0d    25-6382     STARQUEST VENTURES, INC. V. DEXCOM, INC.
2026-05-27   2026-05-27 11:50:06   0d    25-435      SNYDER V. HILL, ET AL.
2026-05-27   2026-05-27 13:00:08   0d    25-1575     POLEJEWSKI V. COUNTY OF CASCADE, ET AL.
2026-05-27   2026-05-27 13:15:04   0d    25-1381     STEPHENSON V. COUNTY OF PLACER
2026-05-27   2026-05-28 13:22:00   1d    25-877      BUNKER, ET AL. V. FRAZIER-KRANE, ET AL.
2026-05-29   2026-05-29 08:50:04   0d    23-2080     LONG V. NOLAND, ET AL.
2026-05-29   2026-05-29 09:00:08   0d    24-5622     SCHWARTZ, ET AL. V. SCHWARTZ, ET AL.
2026-05-29   2026-05-29 09:05:04   0d    24-6954     ACTCA, A MEMBER OF THE ALLIANCE, INC. V. ...
2026-05-29   2026-05-29 09:05:07   0d    25-2335     AHMANN, ET AL. V. WASHINGTON DEPARTMENT ...
2026-05-29   2026-05-29 09:20:08   0d    25-3165     EVANSTON INSURANCE COMPANY V. ROMAN CATHOLIC ...
2026-05-29   2026-05-29 09:30:07   0d    24-5389     ZIEGLER, ET AL. V. WASHINGTON STATE CORR ...
2026-05-29   2026-05-29 09:35:04   0d    25-3880     HALOUSEK V. CALIFORNIA PUBLIC EMPLOYEES'
2026-05-29   2026-05-29 09:45:04   0d    24-5559     MUNENE V. MULLIN, ET AL.
2026-05-29   2026-05-29 09:55:07   0d    25-7474     USA V. KEALOHA
2026-05-29   2026-05-29 10:10:04   0d    25-2341     HILL V. BONTA, ET AL.
```

</details>

## Late-posted items are a second, smaller leak

Separately, when the court posts an item days or months after the filing date and does **not** bump `publish`, the item enters the feed at its stale filing-date position — i.e. at the bottom, behind items we already have — and the same abort guarantees we never reach it. Two confirmed absent from CourtListener (verified with `stat_Published` and `stat_Unpublished`, against a control query that returns correctly):

| docket | filed | posted | lag | in CL? |
|---|---|---|---|---|
| `25-90232` IN RE COMPLAINT OF JUDICIAL MISCONDUCT | 2026-02-02 | 2026-03-31 | 57 days | no |
| `25-5341` GOOGLE LLC V. POINT FINANCIAL, INC. | 2026-05-04 | 2026-05-11 | 7 days | no |

Since 2026-01-01 there were 2 such opinions and 10 such memoranda. 4 of the 12 were rescued only because the court happened to bump `publish` to the posting date, which is why this looks intermittent.

## Proposed fix

Same change as #2119: read the DynamoDB tables behind the site instead of the RSS feeds.

Both tables are readable with the **same anonymous Cognito identity** `ca9` oral arguments already uses (`us-west-2:8d780f3b-d79c-c6c8-1125-e7a905da6b9b`), via `juriscraper.lib.auth_utils.generate_aws_sigv4_headers`:

- `opinions` — 13,810 rows
- `memoranda` — 58,569 rows

Both carry `last_updated` ("YYYY-MM-DD HH:MM:SS"), the court's write timestamp. So:

1. Scan with `FilterExpression` on `last_updated >= :cutoff` over a rolling window.
2. Sort cases by `last_updated` descending, and no-op `_date_sort` so the base class does not re-sort by filing date. This puts every new row above every row already in CL, so the duplicate abort can only fire once the new rows are in.
3. Keep a `publish`-filtered payload for backscrapes, which the feeds cannot support at all today.

The tables also carry metadata the feed does not:

| field | use |
|---|---|
| `case_origin` | `appeal_from_str` — replaces `ca9_p.extract_from_text`'s regex over the PDF text |
| `judge`, `other_judges`, `case_panel` | `author_str` / `panel` |
| `case_type` | Criminal / Agency / etc. |
| `data_source` | CMECF vs ACMS |
| `deleted` | rows to skip |
| `file_name` | PDF path, same value the feed's `<link>` gives |

Worth discussing before a PR: whether to keep `ca9_u` as a subclass of `ca9_p` (the two tables have slightly different columns — `memoranda` has `case_panel`, `opinions` has `judge`/`other_judges`), and how wide the `last_updated` window should be.

## Related

- #2111 — same root cause, `ca9` oral arguments
- #2119 — the oral argument fix
- #2113 — `record_type: "bap"` rows ingested as `ca9`

### PR DESCRIPTION
**Pull request by renatodvc**
## Fixes
Closes #2120 
This fixes missing opinions from ca9_p and ca9_u

## Summary
Refactor ca9_p and ca9_u to extract opinions directly from the DynamoDB available from the source. It processes "cases by last_updated descending, and no-op _date_sort so the base class does not re-sort by filing date" as the issue describes.

Test case fixtures/samples are replaced to accomodate new format.

## AI Disclosure
- [ ] No AI tools were used to create the content of this PR.
- [x] Parts of this PR were created with the help of an AI tool, and I have carefully reviewed all of its content and take full responsibility for it.

### CODE REVIEW

**Comment by renatodvc** [comment]
@flooie

**Comment by grossir** [comment]
@claude review

File: tests/examples/opinions/united_states/ca9_p_example.json [hunk 1]
```diff
@@ -0,0 +1,402 @@
...
+      "S": "ORLONZO HEDRINGTON, ET AL V. USA"
+    },
+    "case_name_lc": {
+      "S": "orlonzo hedrington, et al v. usa"
+    },
+    "case_num": {
+      "S": "23-15191"
+    },
+    "case_origin": {
+      "S": "Sacramento District Court"
```
Comment (grossir) [inline tests/examples/opinions/united_states/ca9_p_example.json:63]: we shouldn't drop "case_origin", we can save it as "lower_court"

Some don't have the "Court" string, so maybe we can filter on that to keep the value

File: tests/examples/opinions/united_states/ca9_p_example_2.json
Comment (grossir) [inline tests/examples/opinions/united_states/ca9_p_example_2.json:1]: this looks like synthetic data. Any reason to use it? We prefer using real data, since it catches general and edge case parsing. Sometimes, we may edit the original input a bit to include an edge case, but we usually don't use synthetic data

File: tests/local/test_CA9Test.py
Comment (grossir) [inline tests/local/test_CA9Test.py:1]: let's not use site specific tests for now

File: juriscraper/opinions/united_states/federal_appellate/ca9_p.py [hunk 2]
```diff
@@ -7,38 +7,321 @@
...
+                ":from_date": {
+                    "S": self.start_date.strftime("%Y-%m-%d 00:00:00")
+                }
+            }
+
+        self.payload = json.dumps(
+            {
+                "TableName": self.table,
+                "ReturnConsumedCapacity": "TOTAL",
+                "FilterExpression": expression,
```
Comment (grossir) [inline juriscraper/opinions/united_states/federal_appellate/ca9_p.py:110]: This may be out of scope for this PR, but we can explore if we can be more gentle with the court's server 

We are using the "scan" operation that pulls all data from their store, and filters it before returning https://github.com/freelawproject/juriscraper/blob/0c0e21ffd55a09b51bf504c2264648ea454886a0/juriscraper/lib/auth_utils.py#L70

but, if available to the anonymous user, we could use the Query endpoint. See [comparison](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/bp-query-scan.html)

Can you check this, and open a follow up issue if you find anything interesting?

Another possible follow up issue is consolidating cognito code for all ca9 scrapers (oral args, opinions, bap9)

**Review by grossir — COMMENTED** [review]
@renatodvc this isn't a complete review, I will leave that for Bill. But some points I found on a initial exploration of the PR

File: tests/examples/opinions/united_states/ca9_p_example_2.json
Comment (flooie) [reply tests/examples/opinions/united_states/ca9_p_example_2.json:1]: @renatodvc is this synthetic data?

File: tests/examples/opinions/united_states/ca9_p_example_2.compare.json [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+  },
+  {
+    "case_dates": "2026-08-28",
+    "case_names": "La International Corp. v. Prestige Brands Holdings, Inc.",
+    "download_urls": "https://cdn.ca9.uscourts.gov/datastore/opinions/2026/08/28/24-5227.pdf",
+    "precedential_statuses": "Published",
+    "blocked_statuses": false,
+    "date_filed_is_approximate": false,
+    "docket_numbers": "24-5227",
+    "lower_courts": "Los Angeles, Central California",
```
Comment (flooie) [inline tests/examples/opinions/united_states/ca9_p_example_2.compare.json]: This is not a court it's a city.  can we figure out why it says this

File: tests/examples/opinions/united_states/ca9_p_example_2.compare.json [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+  },
+  {
+    "case_dates": "2026-08-31",
+    "case_names": "Liu v. Kaiser Permanente Employees Pension Plan for the Permanente Medical Group, Inc.",
+    "download_urls": "https://cdn.ca9.uscourts.gov/datastore/opinions/2026/08/31/24-4303.pdf",
+    "precedential_statuses": "Published",
+    "blocked_statuses": false,
+    "date_filed_is_approximate": false,
+    "docket_numbers": "24-4303",
+    "lower_courts": "San Francisco, Northern California",
```
Comment (flooie) [inline tests/examples/opinions/united_states/ca9_p_example_2.compare.json]: sanme here.  we want the lower court and perhaps its an appeal from a state court located there but thats not a court

File: tests/examples/opinions/united_states/ca9_u_example.compare.json
Comment (flooie) [inline tests/examples/opinions/united_states/ca9_u_example.compare.json:1]: why did we change the file?  did we need an example that was out of order?

File: juriscraper/opinions/united_states/federal_appellate/ca9_p.py [hunk 5]
```diff
@@ -7,38 +7,328 @@
...
+    query_url = "https://dynamodb.us-west-2.amazonaws.com/"
+
+    identity_id = "us-west-2:8d780f3b-d79c-c6c8-1125-e7a905da6b9b"
+    table = "opinions"
+    base_url = "https://cdn.ca9.uscourts.gov/datastore/opinions/"
+    precedential_status = "Published"
+    # Lookback for the regular scrape, in `last_updated` terms.
+    upload_window_days = 7
+    # A Scan returns at most 1MB per request, a full table requires pagination
+    max_pages = 100
```
Comment (flooie) [inline juriscraper/opinions/united_states/federal_appellate/ca9_p.py:51]: is this pulling 100 pages... unless its a back scraper would it be wise to shrink that to just what is needed?

File: juriscraper/opinions/united_states/federal_appellate/ca9_p.py (hunk 2 above)
Comment (flooie) [reply juriscraper/opinions/united_states/federal_appellate/ca9_p.py:110]: I agree.  Also it seems we just published a new PR that handles order for other courts.  

would it be wise to use that here ?

**Review by flooie — CHANGES_REQUESTED** [review]
I dont dislike what you've done here but It would be helpful to have a conversation about the changes before we go this route.

File: tests/examples/opinions/united_states/ca9_p_example.json (hunk 1 above)
Comment (renatodvc) [reply tests/examples/opinions/united_states/ca9_p_example.json:63]: Now using `lower_court`, will add the "Court" filter as the field can contain just locations

File: tests/examples/opinions/united_states/ca9_p_example_2.json
Comment (renatodvc) [reply tests/examples/opinions/united_states/ca9_p_example_2.json]: It was synthetic for the test case, now it's real from the scraper

File: tests/local/test_CA9Test.py
Comment (renatodvc) [reply tests/local/test_CA9Test.py:1]: removed the test case

File: tests/examples/opinions/united_states/ca9_p_example_2.compare.json (hunk 3 above)
Comment (renatodvc) [reply tests/examples/opinions/united_states/ca9_p_example_2.compare.json]: This is a misinterpretation from my side. I adopted `case_origin` for `lower_courts` as suggested [here](https://github.com/freelawproject/juriscraper/pull/2159#discussion_r3896503365). 

I assumed these refered to _local courts_ (aka lower), but it seems that is only true if they distinctly name the "Court" word in it. Is there any other words I might need to keep an eye for?

---

To clarify: "Los Angeles, Central California" came from the `case_origin` record itself.

File: tests/examples/opinions/united_states/ca9_p_example_2.json
Comment (renatodvc) [reply tests/examples/opinions/united_states/ca9_p_example_2.json:1]: It was synthetic data at the time of the first comment. It was replaced in commit 47cf1fcbee137b1d0978c87a7b877da3de3ee6a1 for real data

File: tests/examples/opinions/united_states/ca9_u_example.compare.json
Comment (renatodvc) [reply tests/examples/opinions/united_states/ca9_u_example.compare.json:1]: It changed because the sample input is no longer the 100 entry RSS feed, and it gained new data such as `judges`, `lower_courts` and `nature_of_suit` from the table's records.

File: juriscraper/opinions/united_states/federal_appellate/ca9_p.py (hunk 2 above)
Comment (renatodvc) [reply juriscraper/opinions/united_states/federal_appellate/ca9_p.py:110]: I opened follow up issues #2163  and #2164  for them.

> I agree. Also it seems we just published a new PR that handles order for other courts.

Do you mean #2162 or another PR?

File: tests/examples/opinions/united_states/ca9_p_example_2.compare.json (hunk 4 above)
Comment (renatodvc) [reply tests/examples/opinions/united_states/ca9_p_example_2.compare.json]: Answered [here](https://github.com/freelawproject/juriscraper/pull/2159#discussion_r3928385757)

File: juriscraper/opinions/united_states/federal_appellate/ca9_p.py (hunk 5 above)
Comment (renatodvc) [reply juriscraper/opinions/united_states/federal_appellate/ca9_p.py:51]: No, it act as a ceiling, just like we do in ca9's `OralArgumentSiteLinear`. 

What actually controls the pagination is the payload's not including a `LastEvaluatedKey`. 

However, as @grossir mentioned [here](https://github.com/freelawproject/juriscraper/pull/2159#discussion_r3896628272), since we are using the `Scan` endpoint, we inevitably force them to read all rows of the DB. 

The only way around it is moving to Query endpoint. #2163
